planner/core/task.go

```diff
@@ -431,20 +454,36 @@ func (p *PhysicalHashJoin) GetCost(lCnt, rCnt float64) float64 {
 	// Cost of quering hash table is cheap actually, so we just compute the cost of
 	// evaluating `OtherConditions` and joining row pairs.
 	probeCost := numPairs * sessVars.CPUFactor
+	probeDiskCost := numPairs * sessVars.DiskFactor * rowSize
 	// Cost of evaluating outer filter.
 	if len(p.LeftConditions)+len(p.RightConditions) > 0 {
 		// Input outer count for the above compution should be adjusted by selectionFactor.
 		probeCost *= selectionFactor
-		probeCost += outerCnt * sessVars.CPUFactor
+		probeCost += probeCnt * sessVars.CPUFactor
+		probeDiskCost *= selectionFactor
+		probeDiskCost += probeCnt * sessVars.DiskFactor * rowSize
 	}
+	diskCost += probeDiskCost
 	probeCost /= float64(p.Concurrency)
 	// Cost of additional concurrent goroutines.
 	cpuCost += probeCost + float64(p.Concurrency+1)*sessVars.ConcurrencyFactor
 	// Cost of traveling the hash table to resolve missing matched cases when building the hash table from the outer table
 	if p.UseOuterToBuild {
-		cpuCost += innerCnt * sessVars.CPUFactor / float64(p.Concurrency)
+		if spill {
+			// It runs in sequence when build data is on disk. See handleUnmatchedRowsFromHashTableInDisk
+			cpuCost += buildCnt * sessVars.CPUFactor
+		} else {
+			cpuCost += buildCnt * sessVars.CPUFactor / float64(p.Concurrency)
+		}
+		diskCost += buildCnt * sessVars.DiskFactor * rowSize
 	}
-	return cpuCost + memoryCost
+
+	if spill {
+		memoryCost *= float64(memQuota) / (rowSize * buildCnt)
```

Comment:
What does this mean?

Reply:
Does these code can describe the meaning?
```go
memUsed := float64(memQuota)
originMemUsed := rowSize * buildCnt
factor := memUsed / originMemUsed
memoryCost *= factor
```

Reply:
Got you, I misunderstood it before.
I think `memoryCost *= float64(memQuota) / (rowSize * buildCnt)` is ok.